SmartDeviceLink/SDLMenuManager.m

```diff
@@ -0,0 +1,400 @@
+
+        weakSelf.oldMenuCells = weakSelf.menuCells;
+
+        [weakSelf.connectionManager sendRequests:subMenuCommands progressHandler:^(__kindof SDLRPCRequest * _Nonnull request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error, float percentComplete) {
+            if (error != nil) {
+                errors[request] = error;
+            }
+        } completionHandler:^(BOOL success) {
+            if (!success) {
+                SDLLogE(@"Failed to send sub menu commands: %@", errors);
+                completionHandler([NSError sdl_menuManager_failedToUpdateWithDictionary:errors]);
+
+                return;
+            }
+
+            SDLLogD(@"Finished updating menu");
+            completionHandler(nil);
+        }];
+    }];
+}
+
+#pragma mark - Setters
+
+- (void)setMenuCells:(NSArray<SDLMenuCell *> *)menuCells {
+    if (self.currentLevel == nil || [self.currentLevel isEqualToString:SDLHMILevelNone]) {
+        _waitingOnHMILevelUpdate = YES;
+        _menuCells = menuCells;
+        return;
+    }
+
+    // Check for duplicate titles
+    NSMutableSet *titleCheckSet = [NSMutableSet set];
+    for (SDLMenuCell *cell in menuCells) {
+        [titleCheckSet addObject:cell.title];
+    }
+    if (titleCheckSet.count != menuCells.count) {
+        SDLLogE(@"Not all cell titles are unique. The menu will not be set.");
+        return;
+    }
+
+    // Set the ids
+    self.lastMenuId = MenuCellIdMin;
+    [self sdl_updateIdsOnMenuCells:menuCells parentId:ParentIdNotFound];
+
+    _oldMenuCells = _menuCells;
+    _menuCells = menuCells;
+
+    // Upload the artworks
+    NSArray<SDLArtwork *> *artworksToBeUploaded = [self sdl_findAllArtworksToBeUploadedFromCells:self.menuCells];
+    if (artworksToBeUploaded.count > 0) {
+        [self.fileManager uploadArtworks:artworksToBeUploaded completionHandler:^(NSArray<NSString *> * _Nonnull artworkNames, NSError * _Nullable error) {
+            if (error != nil) {
+                SDLLogE(@"Error uploading menu artworks: %@", error);
+            }
+
+            SDLLogD(@"Menu artworks uploaded");
+            [self sdl_updateWithCompletionHandler:nil];
+        }];
+    }
+
+    [self sdl_updateWithCompletionHandler:nil];
+}
+
+#pragma mark - Helpers
+
+#pragma mark Artworks
+
+- (NSArray<SDLArtwork *> *)sdl_findAllArtworksToBeUploadedFromCells:(NSArray<SDLMenuCell *> *)cells {
+    if (![self.displayCapabilities hasImageFieldOfName:SDLImageFieldNameCommandIcon]) {
+        return @[];
+    }
+
+    NSMutableArray<SDLArtwork *> *mutableArtworks = [NSMutableArray array];
+    for (SDLMenuCell *cell in cells) {
+        if (cell.icon != nil && ![self.fileManager hasUploadedFile:cell.icon]) {
+            [mutableArtworks addObject:cell.icon];
+        }
+
+        if (cell.subCells.count > 0) {
+            [mutableArtworks addObjectsFromArray:[self sdl_findAllArtworksToBeUploadedFromCells:cell.subCells]];
+        }
+    }
+
+    return [mutableArtworks copy];
```

Comment:
If a choice set with 100 items has all the same icon, the same icon will get sent to Core 100 times.